### PR TITLE
Add more descriptive error logging to init info and resource intializations

### PR DESCRIPTION
**Pull request by sndels**

### CODE REVIEW

File: include/Core/Binary_Init_File.hpp [hunk 1]
```diff
@@ -14,28 +15,49 @@ namespace fug {
...
         }
-        if (fseek(fp, 0, SEEK_END) < 0) {
+        // Get size of file
+        int err = 0;
+        if ((err = fseek(fp, 0, SEEK_END)) < 0) {
             fclose(fp);
-            throw "Binary: fseek failed";
+            FUG_LOG(LogLevel::Error)("BinaryInitInfo_File: fseek SEEK_END failed\n");
+            FUG_LOG(LogLevel::Error)("File: %s\n", initInfo.fileName.c_str());
+            FUG_LOG(LogLevel::Error)("Error: %i\n", err);
```
Comment (tkln) [inline include/Core/Binary_Init_File.hpp]: This print does not have any value in itself as err is always going to be -1 if this branch is taken. I suggest either removing this or printing strerror(errno).

File: include/Core/Binary_Init_File.hpp [hunk 2]
```diff
@@ -14,28 +15,49 @@ namespace fug {
...
+            throw;
         }
         size_ = ftell(fp);
-        if (fseek(fp, 0, SEEK_SET) < 0) {
+        // Reset file pointer to start of file
+        if ((err = fseek(fp, 0, SEEK_SET)) < 0) {
             fclose(fp);
-            throw "Binary: fseek failed";
+            FUG_LOG(LogLevel::Error)("BinaryInitInfo_File: fseek SEEK_SET failed\n");
+            FUG_LOG(LogLevel::Error)("File: %s\n" ,initInfo.fileName.c_str());
```
Comment (tkln) [inline include/Core/Binary_Init_File.hpp]: Space on the wrong side of a comma.

File: include/Core/Binary_Init_File.hpp [hunk 3]
```diff
@@ -14,28 +15,49 @@ namespace fug {
...
+            FUG_LOG(LogLevel::Error)("Error: %i\n", err);
```
Comment (tkln) [inline include/Core/Binary_Init_File.hpp]: Same as on line 40.

File: include/Core/JsonUtility.hpp [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+#define FUG_CORE_JSON_UTILITY_HPP
+
+#include <unordered_set>
+#include <json/json.hpp>
+
+namespace fug {
+
+// Checks that given json doesn't have invalid keys and contains all required keys
+// Potential errors are printed to FUG_LOG(LogLevel::Error) and result in a throw
+void checkJsonValidity(const std::string& type, const nlohmann::json& json,
```
Comment (tkln) [inline include/Core/JsonUtility.hpp]: Based on the name, I would have assumed this to return something. This seems to something more of an assert.

File: include/Core/Binary_Init_File.hpp [hunk 5]
```diff
@@ -14,28 +15,49 @@ namespace fug {
...
         initInfo.fileName = std::string(FUG_RES_DIRECTORY) + fileName;
     }
 
     FUG_RESOURCE_INIT(Binary, BinaryInitInfo_File) {
+        // Open file
         FILE *fp = fopen(initInfo.fileName.c_str(), "rb");
         if (!fp) {
-            throw "Binary: open failed";
+            FUG_LOG(LogLevel::Error)("BinaryInitInfo_File: fopen failed\n");
+            FUG_LOG(LogLevel::Error)("File: %s\n", initInfo.fileName.c_str());
```
Comment (tkln) [inline include/Core/Binary_Init_File.hpp:32]: Printing strerr(errno) could be useful.

File: include/Core/Text_Init_File.hpp [hunk 6]
```diff
@@ -13,28 +14,49 @@ namespace fug {
...
         }
         size_ = ftell(fp);
-        if (fseek(fp, 0, SEEK_SET) < 0) {
+        // Reset file pointer to start of file
+        if ((err = fseek(fp, 0, SEEK_SET)) < 0) {
             fclose(fp);
-            throw "Text: fseek failed";
+            FUG_LOG(LogLevel::Error)("TextInitInfo_File: fseek SEEK_SET failed\n");
+            FUG_LOG(LogLevel::Error)("File: %s\n" ,initInfo.fileName.c_str());
+            FUG_LOG(LogLevel::Error)("Error: %i\n", err);
```
Comment (tkln) [inline include/Core/Text_Init_File.hpp]: Ditto

File: include/Core/Text_Init_File.hpp [hunk 7]
```diff
@@ -13,28 +14,49 @@ namespace fug {
...
+            throw;
         }
         size_ = ftell(fp);
-        if (fseek(fp, 0, SEEK_SET) < 0) {
+        // Reset file pointer to start of file
+        if ((err = fseek(fp, 0, SEEK_SET)) < 0) {
             fclose(fp);
-            throw "Text: fseek failed";
+            FUG_LOG(LogLevel::Error)("TextInitInfo_File: fseek SEEK_SET failed\n");
+            FUG_LOG(LogLevel::Error)("File: %s\n" ,initInfo.fileName.c_str());
```
Comment (tkln) [inline include/Core/Text_Init_File.hpp]: Ditto

File: include/Core/Text_Init_File.hpp [hunk 8]
```diff
@@ -13,28 +14,49 @@ namespace fug {
...
         }
-        if (fseek(fp, 0, SEEK_END) < 0) {
+        // Get size of file
+        int err = 0;
+        if ((err = fseek(fp, 0, SEEK_END)) < 0) {
             fclose(fp);
-            throw "Text: fseek failed";
+            FUG_LOG(LogLevel::Error)("TextInitInfo_File: fseek SEEK_END failed\n");
+            FUG_LOG(LogLevel::Error)("File: %s\n", initInfo.fileName.c_str());
+            FUG_LOG(LogLevel::Error)("Error: %i\n", err);
```
Comment (tkln) [inline include/Core/Text_Init_File.hpp]: Ditto

File: include/Core/Text_Init_File.hpp [hunk 9]
```diff
@@ -13,28 +14,49 @@ namespace fug {
...
         initInfo.fileName = std::string(FUG_RES_DIRECTORY) + fileName;
     }
 
     FUG_RESOURCE_INIT(Text, TextInitInfo_File) {
+        // Open file
         FILE *fp = fopen(initInfo.fileName.c_str(), "rb");
         if (!fp) {
-            throw "Text: open failed";
+            FUG_LOG(LogLevel::Error)("TextInitInfo_File: fopen failed\n");
+            FUG_LOG(LogLevel::Error)("File: %s\n", initInfo.fileName.c_str());
```
Comment (tkln) [inline include/Core/Text_Init_File.hpp:30]: Same as Binary_Init_File.hpp:31

**Review by tkln — APPROVED** [review]

File: include/Core/Binary_Init_File.hpp (hunk 5 above)
Comment (sndels) [reply include/Core/Binary_Init_File.hpp:32]: check

File: include/Core/Text_Init_File.hpp (hunk 9 above)
Comment (sndels) [reply include/Core/Text_Init_File.hpp:30]: check

**Pull request merged**